bazaarvoice/emodb
# Added caching of oldest slab to readAll() queries

## Github Issue #

## What Are We Doing Here?

Polling from the event store uses a method in `AstyanaxEventReaderDAO` called `readNewer()`.  This method uses a cache of the oldest slab found per channel in the manifest called oldestSlab.  When the oldest slab from a previous query is known future queries for up to the next 10 seconds include a clause "where slabID > oldestSlab - 1 minute".  This allows manifest wide rows with many initial tombstones to bypass scanning those tombstones to once every 10 seconds.

The peek method, `readAll()`, does not use any such caching and reads from the start of the wide row on every request.  This shouldn't be a big deal because peeks are far less frequent than polls.  However, data center replication (the means for fanning out databus updates in us-east-1 to eu-west-1 and vice versa) uses peeks, not polls, and evidence shows that some partitions are much slower to poll as a result.

This PR extends the same logic used by `readNewer()` to `readAll()`.  Note that `readAll()` takes an additional `Date` which can be used to filter events only newer than a particular time.  However, this is done by filtering the slabs returned from the manifest and does not effect the manifest query itself, so even with filtering by date the query always returns all manifest entries so the oldest can safely be cached.

Additionally this PR removes the `ConsistencyLevel` parameter from `readAll()` since in all calls to the method the value is always `ConsistencyLevel.CL_LOCAL_QUORUM`.

## How to Test and Verify

This change only affects performance of the manifest query when polling and peeking.  As such it is difficult to directly test.  Rather the best way to test is through full regression to ensure no events are lost as a result of this change.

## Risk

Medium.  Although this is a relatively minor change it is in the heart of queries used for polling and peeking the databus and queues, so this raises the risk associated.

### Level 

`Medium`

### Required Testing

`Regression`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
